library(shiny)
library(sangerseqR)
library(xlsx)
source("procAbi.R")
source("helpers.R")
source("samples.R")


shinyServer(function(input,output,session) {
    #options(shiny.reactlog=TRUE)
    g_calls                 <- NULL             #annotated basecall data
    #makeReactiveBinding("g_calls")
    g_intens                <- NULL             #intensities file
    g_intens_rev            <- NULL             #optional reverse intensities file
    g_single_rev            <- NULL
    g_intrexdat             <- NULL             #intrex data used in graphs
    g_glassed_ref           <- NULL
    g_glassed_cod           <- NULL
    g_glassed_snp           <- NULL
    g_choices               <- NULL
    g_noisy_neighbors       <- NULL
    g_view                  <- NULL
    g_selected              <- NULL
    g_selected_goto_index   <- 0
    g_max_y                 <- NULL
    g_hetero_calls          <- 0
    g_hetero_indel_pid      <- 0
    g_hetero_ins_tab        <- NULL
    g_hetero_del_tab        <- NULL
    g_expected_het_indel    <- NULL
    g_minor_het_insertions  <- NULL
    load_id                 <- NULL
    g_stored_het_indels     <- list()
    files_info              <- ""
    g_indels_present        <- FALSE
    g_qual_present          <- FALSE
    g_brush_fwd             <- 0
    g_brush_rev             <- 0
    g_not_loaded            <- ""
    g_reactval              <- reactiveValues()
    g_reactval$updateVar    <- 0
    g_refs_avail            <<- c("-","TP53","NOTCH1","ATM","CALR")
    g_files                 <- data.table(FWD_name=c("TP53 ; fwd ; low freq w frameshift"),
                                          FWD_file=c("data/abis/eric/3low_freq_fsF.ab1"),
                                          REV_name=c("TP53 ; rev ; low freq w frameshift"),
                                          REV_file=c("data/abis/eric/3low_freq_fsR.ab1"),
                                          REF=c("TP53"),
                                          mut_min=20,qual_thres_to_call=0,s2n_min=2,show_calls_checkbox=F,join_traces_checkbox=F,max_y_p=100,opacity=0,incorporate_checkbox=F,loaded=F,
                                          calls = "",
                                          status="new",
                                          id=1,
                                          brush_fwd = 0,
                                          brush_rev = 0
                                )

#     get_file <- reactive({
#        # if (!is.null(input$select_file)) return(input$select_file$datapath)
#         if (!is.null(input$select_file)) return(input$select_file)
#         else return(NULL)
#     })
    ex <- NULL
    btn_counter <- 0

    #SAMPLE BROWSER STUFF

    output$samples_table <- DT::renderDataTable({
        loadSamples()
        goRef_handler()
        goDelete_sample_handler()
        goSwap_sample_handler()
        goChangeRev_handler()
        loading_processed_files()
        input$goLock

        disabled <- rep(FALSE,nrow(g_files))
        disabled[1] <- FALSE
        g_files[,id:= 1:nrow(g_files)]
        g_files <<- g_files
        add_load_buttons     <- shinyInput(actionButton, 1:nrow(g_files), 'loadSample_', label = NULL, onclick = 'Shiny.onInputChange(\"goLoadSamples\",  this.id + (Math.random()/10))',ico=rep("play",nrow(g_files)),class="btn btn-info" )
        add_delete_buttons   <- shinyInput(actionButton, 1:nrow(g_files), 'delSample_',  label = NULL, onclick = 'Shiny.onInputChange(\"goDeleteSamples\",  this.id)',ico=rep("close",nrow(g_files)) ,dsbl = disabled,class="btn dlt_btn")
        add_swap_buttons     <- shinyInput(actionButton, 1:nrow(g_files), 'swapSample_', label = NULL, onclick = 'Shiny.onInputChange(\"goSwapSamples\",  this.id + (Math.random()/10))',ico=rep("exchange",nrow(g_files))  ,dsbl = disabled)
        #add_reference_dropdown <- shinyInput(selectizeInput, 1:nrow(g_files), 'selectInput_',choices=c("TP53","NOTCH1","ATM"),onchange = 'Shiny.onInputChange(\"goChangeSamples\",  this.id + (Math.random()/10))',label=NULL,width="100px") #selected = ref
        add_reference_dropdown <- shinyInput(selectInput, 1:nrow(g_files), 'selectGene_',choices=g_refs_avail, selected = g_files[,REF],width="80px")
        #add_reference_dropdown <- shinyInput(selectInput, 1:nrow(g_files), 'selectInput_',choices=c("TP53","NOTCH1","ATM","FOUR","FIVE","SIX"),label=NULL) #selected = ref
        add_reverse_dropdown <- shinyInputRev(selectInput,1:nrow(g_files),'chooseRev_',g_files,width="240px")
        #out<-cbind(g_files[,list("forward"=FWD_name)],"swap"=add_swap_buttons,g_files[,list("reverse"=REV_name)],"reference"=add_reference_dropdown,delete=add_delete_buttons,load=add_load_buttons)
        out<-cbind(" "=add_delete_buttons,g_files[,list("forward"=FWD_name)],"swap"=add_swap_buttons,"<div title='Use dropdown menu to pair or unpair samples. Only unpaired reverse files appear here.' >reverse [?]</div>"=add_reverse_dropdown,"<div title='' >reference</div>"=add_reference_dropdown," "=add_load_buttons,g_files[,list("<div title='Confirmed (locked) variants appear here.'>status [?]</div>"=status)])
        table_out <- DT::datatable(out,escape=FALSE,
                                   selection = "none",
                                   style = "bootstrap",
                                   #id = "samplesdt",
                                   class="compact",
                                   options=list("ordering"=FALSE,"paging"=FALSE,"searching"=FALSE,"autoWidth"=FALSE,"bInfo"=FALSE,"id"="samplesdt",
                                                initComplete = JS('function(setting, json) {
                                                                        $(\'[id*="selectGene"]\').change(function() {
                                                                                var sel = $(this).find(":selected").text();
                                                                                Shiny.onInputChange("goChangeRef",{id: this.id,
                                                                                                                   gene: sel});
                                                                        });
                                                                        $(\'[id*="chooseRev"]\').change(function() {
                                                                                var sel = $(this).find(":selected").text();
                                                                                Shiny.onInputChange("goChangeRev",{id: this.id,
                                                                                                                   name: sel});
                                                                                /*alert("changing rev "+this.id+" to "+ sel);*/
                                                                        });
                                                                  }')
                                                )
                                   )
    })

    #Handlers for the Sample Browser

    loadSamples <- reactive({
        if(!is.null(input$browser_files)){
            g_not_loaded <- ""
            loaded <- ""
            error = ""
            ret <- samples_load(input$browser_files,output,g_files,input$alignTo)
            g_files <<- rbind(g_files[,!c("id"),with=FALSE],ret$loaded)
            g_files[,id:= 1:nrow(g_files)]
            g_files <<- g_files
            if(length(ret$not_loaded)>1){
                output$files <- renderPrint(paste0("<pre>Unable to load the following files: ",paste(ret$not_loaded[2:length(ret$not_loaded)],collapse="<br>"),"</pre>" ))
            }
        }
    })

    goChangeRev_handler <- reactive({
        if(!is.null(input$goChangeRev)){
            pos_at <- as.numeric(strsplit(input$goChangeRev$id,"_")[[1]][2])
            name <- as.character(input$goChangeRev$name)
            if(name=="-"){  #split
                rev_name <- g_files[pos_at]$REV_name
                rev_file <- g_files[pos_at]$REV_file
                ref      <- g_files[pos_at]$REF

                g_files[pos_at]$REV_name <- "-"
                g_files[pos_at]$REV_file <- "-"
                g_files[pos_at]$calls <- ""
                g_files[pos_at]$loaded <- FALSE
                g_files[pos_at]$status <- "new"
                g_files[pos_at]$brush_rev <- 0
                g_files[pos_at]$calls = ""
                g_files <<- rbind(g_files,c(list(FWD_name="-",FWD_file="-",REV_name=rev_name,REV_file=rev_file,REF=ref,id=nrow(g_files)+1),mut_min=20,qual_thres_to_call=0,s2n_min=2,show_calls_checkbox=F,join_traces_checkbox=F,max_y_p=100,opacity=0,incorporate_checkbox=F,calls = "",loaded= FALSE, status = "new",brush_fwd = 0, brush_rev = 0))
            }else{          #combine
                setkey(g_files,id)
                rev_name <- name
                rm_id <- g_files[g_files[,REV_name == name]]$id
                rev_file <- g_files[g_files[,REV_name == name]]$REV_file
                pos_at <- as.numeric(strsplit(input$goChangeRev$id,"_")[[1]][2])
                g_files[pos_at]$REV_name <- rev_name
                g_files[pos_at]$REV_file <- rev_file
                g_files<<-g_files[!g_files[id==rm_id]]

            }
        }
    })
    goRef_handler <- reactive({
        if(!is.null(input$goChangeRef)){
            ref_id <- as.numeric(strsplit(input$goChangeRef$id,"_")[[1]][2])
            g_files[ref_id]$REF <-  as.character(input$goChangeRef$gene)
            g_files <<- g_files
        }
    })

    goDelete_sample_handler <- reactive({
        if(!is.null(input$goDeleteSamples)){
            isolate({
                delete_id <- as.numeric(strsplit(input$goDeleteSamples, "_")[[1]][2])
                #g_files <<- g_files[-delete_id]
                g_files <<- g_files[!g_files[,id==delete_id]]
                #js$delRow(delete_id)
            })
        }
    })

    goSwap_sample_handler <- reactive({
        input$goSwapSamples
        if(!is.null(input$goSwapSamples)){
            isolate({
                swap_id <- floor(as.numeric(strsplit(input$goSwapSamples, "_")[[1]][2]))/10
                swp_name <- g_files[id==swap_id]$FWD_name
                swp_file <- g_files[id==swap_id]$FWD_file
                g_files[id==swap_id]$FWD_name <- g_files[id==swap_id]$REV_name
                g_files[id==swap_id]$FWD_file <- g_files[id==swap_id]$REV_file
                g_files[id==swap_id]$REV_file <- swp_file
                g_files[id==swap_id]$REV_name <- swp_name
                g_files <<- g_files
                #js$swapRow(swap_id)

                updateSelectizeInput(
                    session, 'selectizeInput_1',server = FALSE,
                    options = list ("maxInput" = 5)
                )
            })
        }
    })

    #save settings in sample browser
    controls_listener <- observe({
        input$mut_min
        input$qual_thres_to_call
        input$max_y_p
        input$s2n_min
        input$show_calls_checkbox
        input$join_traces_checkbox
        input$max_y_p
        input$opacity
        input$incorporate_checkbox
        g_reactval$updateVar

        if(nrow(g_files[loaded==TRUE,]) ==1){
            g_files <<- g_files[loaded==TRUE,`:=`(mut_min=input$mut_min,
                                      qual_thres_to_call=input$qual_thres_to_call,
                                      s2n_min=input$s2n_min,
                                      show_calls_checkbox=input$show_calls_checkbox,
                                      join_traces_checkbox=input$join_traces_checkbox,
                                      max_y_p=input$max_y_p,
                                      opacity=input$opacity,
                                      incorporate_checkbox=input$incorporate_checkbox,
                                      brush_fwd=g_brush_fwd,
                                      brush_rev=g_brush_rev)]
        }


    })
#    change_reference <- observe ({
#        input$gene_of_interest
#        isolate({
#            g_glassed_ref <<- paste("data/refs/",input$gene_of_interest,".glassed.intrex.fasta",sep="")
#            g_glassed_cod <<- paste("data/refs/",input$gene_of_interest,".glassed.codons.rdata",sep="")
#            #g_files <<- paste0("Reference changed to ", input$gene_of_interest)
#            #if(class(loading_processed_files())[1] != "my_UI_exception"){
#                #output$files      <-  renderPrint({cat(g_files)})
#            #}
#        })
#        #return(list(glassed_ref=glassed_ref,glassed_cod=glassed_cod))
#    })

    loading_processed_files <- reactive ({
        #if (input$ex_btn[1] - btn_counter){
        #    btn_counter <<- input$ex_btn[1]
        #    ex <- c("data/abis/eric/3low_freq_fsF.ab1",
        #            "data/abis/eric/3low_freq_fsR.ab1")
        #}
        calls <- structure("error_reading_Rbin",class = "my_UI_exception")

        if(!is.null(input$goLoadSamples)){
            isolate({
                load_id <- floor(as.numeric(strsplit(input$goLoadSamples, "_")[[1]][2]))/10
                if(nrow(g_files[loaded==TRUE,])==1){
                    tmpfile <- tempfile("calls")
                    save(g_calls,file=tmpfile)
                    g_files[loaded==TRUE,calls:=tmpfile]
                }
                g_files[,loaded:=F]
                g_files[load_id,loaded:=T]
                if(g_files[loaded==T,]$calls == ""){
                    g_files[load_id,status:="viewed"]
                }
                updateSliderInput(session,'mut_min',value=g_files[loaded==T,]$mut_min)
                updateSliderInput(session,'qual_thres_to_call',value=g_files[loaded==T,]$qual_thres_to_call)
                updateSliderInput(session,'s2n_min',value=g_files[loaded==T,]$s2n_min)
                updateCheckboxInput(session,'show_calls_checkbox',value=g_files[loaded==T,]$show_calls_checkbox)
                updateCheckboxInput(session,'join_traces_checkbox',value=g_files[loaded==T,]$join_traces_checkbox)
                updateSliderInput(session,'max_y_p',value=g_files[loaded==T,]$max_y_p)
                updateSliderInput(session,'opacity',value=g_files[loaded==T,]$opacity)
                updateCheckboxInput(session,'incorporate_checkbox',value=g_files[loaded==T,]$incorporate_checkbox)
                #updateCheckboxInput(session,'incorporate_checkbox',value=FALSE)
                g_files<<-g_files
            })
        }
        single_rev <- FALSE
        #if(!is.null(input$select_file) || !is.null(ex)) {
        if(!is.null(load_id) || !is.null(ex)) {
            fwd_file      <- g_files[load_id]$FWD_file
            fwd_file_name <- g_files[load_id]$FWD_name
            rev_file      <- g_files[load_id]$REV_file
            rev_file_name <- g_files[load_id]$REV_name
            ref           <- g_files[load_id]$REF

            if(!is.null(ex)){
                fwd_file <- fwd_file_name <- ex[1]
                rev_file <- rev_file_name <- ex[2]
                ref = "TP53"
            }
            g_glassed_ref <<- paste("data/refs/",ref,".glassed.intrex.fasta",sep="")
            g_glassed_cod <<- paste("data/refs/",ref,".glassed.codons.rdata",sep="")
            snp_file      <-  paste("data/refs/",ref,".dbSNP.tsv",sep="")
            if(file.exists(snp_file)){
                g_glassed_snp <<- fread(snp_file)
            }else{
                g_glassed_snp <<- NULL
            }

            if (fwd_file_name == "-"){
                single_rev <- TRUE
                fwd_file <- rev_file
                rev_file <- NULL
            }
            else
                single_rev <- FALSE
            if (rev_file_name == "-")
                rev_file <- NULL

            g_single_rev <<- single_rev
          #  file <- input$select_file$datapath
          #  name <- input$select_file$name
          #  #input$gene_of_interest
          #  full_name <- input$select_file$name
          #  single_rev <- FALSE
          #
          #  #getting rid of the date in names
          #  for(i in 1:length(name)){
          #      name[i] <- gsub("_[12][09][0-9][0-9]-[0-1][0-9]-[0123][0-9]_[0-1][0-9]-[0-5][0-9]-[0-5][0-9]","",name[i])
          #      name[i] <- gsub(".abi",".ab1",name[i])
          #  }
          #  #if multiple files uploaded we use the first to
          #  #check if we can distinguish forward and reverse
          #  #otherwise we only take the first file
          #  base <- ""
          #  if(length(name)>=2){
          #      if(gsub("*F.ab1","F",name[1])==gsub("*R.ab1","F",name[2])){
          #          fwd_file <- input$select_file$datapath[1]
          #          fwd_file_name <- full_name[1]
          #          rev_file <- input$select_file$datapath[2]
          #          rev_file_name <- full_name[2]
          #      }else if(gsub("*R.ab1","R",name[1])==gsub("*F.ab1","R",name[2])){
          #          fwd_file <- input$select_file$datapath[2]
          #          fwd_file_name <- full_name[2]
          #          rev_file <- input$select_file$datapath[1]
          #          rev_file_name <- full_name[1]
          #      }else{
          #          fwd_file <- input$select_file$datapath[1]
          #          fwd_file_name <- full_name[1]
          #          rev_file <- NULL
          #          rev_file_name <- "-"
          #      }
          #  }else if(is.null(ex)){ #only one file selected
          #      fwd_file <- input$select_file$datapath
          #      fwd_file_name <- full_name
          #      rev_file <- NULL
          #      rev_file_name <- "-"
          #      ref <- read.fasta(paste0("data/refs/",input$gene_of_interest,".glassed.intrex.fasta"))
          #      re
          #      sm <<- matrix(c(1 ,-1 ,-1 ,-1 ,-1 ,0.5 ,0.5 ,-1 ,-1 ,0.5 ,-1 ,0.1 ,0.1 ,0.1 ,0 ,-1 ,1 ,-1 ,-1 ,-1 ,0.5 ,-1 ,0.5 ,0.5 ,-1 ,0.1 ,-1 ,0.1 ,0.1 ,0 ,-1 ,-1 ,1 ,-1 ,0.5 ,-1 ,0.5 ,-1 ,0.5 ,-1 ,0.1 ,0.1 ,-1 ,0.1 ,0 ,-1 ,-1 ,-1 ,1 ,0.5 ,-1 ,-1 ,0.5 ,-1 ,0.5 ,0.1 ,0.1 ,0.1 ,-1 ,0 ,-1 ,-1 ,0.5 ,0.5 ,0.1 ,-1 ,0 ,0 ,0 ,0 ,0.1 ,0.1 ,-0.1 ,-0.1 ,0.1 ,0.5 ,0.5 ,-1 ,-1 ,-1 ,0.1 ,0 ,0 ,0 ,0 ,-0.1 ,-0.1 ,0.1 ,0.1 ,0.1 ,0.5 ,-1 ,0.5 ,-1 ,0 ,0 ,0.1 ,-1 ,0 ,0 ,-0.1 ,0.1 ,-0.1 ,0.1 ,0.1 ,-1 ,0.5 ,-1 ,0.5 ,0 ,0 ,-1 ,0.1 ,0 ,0 ,0.1 ,-0.1 ,0.1 ,-0.1 ,0.1 ,-1 ,0.5 ,0.5 ,-1 ,0 ,0 ,0 ,0 ,0.1 ,-1 ,0.1 ,-0.1 ,-0.1 ,0.1 ,0.1 ,0.5 ,-1 ,-1 ,0.5 ,0 ,0 ,0 ,0 ,-1 ,0.1 ,-0.1 ,0.1 ,0.1 ,-0.1 ,0.1 ,-1 ,0.1 ,0.1 ,0.1 ,0.1 ,-0.1 ,-0.1 ,0.1 ,0.1 ,-0.1 ,0.1 ,0 ,0 ,0 ,0.1 ,0.1 ,-1 ,0.1 ,0.1 ,0.1 ,-0.1 ,0.1 ,-0.1 ,-0.1 ,0.1 ,0 ,0.1 ,0 ,0 ,0.1 ,0.1 ,0.1 ,-1 ,0.1 ,-0.1 ,0.1 ,-0.1 ,0.1 ,-0.1 ,0.1 ,0 ,0 ,0.1 ,0 ,0.1 ,0.1 ,0.1 ,0.1 ,-1 ,-0.1 ,0.1 ,0.1 ,-0.1 ,0.1 ,-0.1 ,0 ,0 ,0 ,0.1 ,0.1 ,0 ,0 ,0 ,0 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1 ,0.1),15,15,dimnames = list(c("A","T","G","C","S","W","R","Y","K","M","B","V","H","D","N"),c("A","T","G","C","S","W","R","Y","K","M","B","V","H","D","N")))
          #      seq<-DNAString(sangerseqR::read.abif(fwd_file)@data$PBAS.2)
          #      pa <- pairwiseAlignment(pattern = ref, subject = seq,type = "local",substitutionMatrix = sm,gapOpening = -6, gapExtension = -1)
          #      score_fwd <- pa@score
          #      pa <- pairwiseAlignment(pattern = ref, subject = reverseComplement(seq),type = "local",substitutionMatrix = sm,gapOpening = -6, gapExtension = -1)
          #      score_rev <- pa@score
          #      if(score_rev>score_fwd){
          #          single_rev <- TRUE
          #      }
          #      #if(!is.null(name)){
          #      #    base <- sapply(strsplit(basename(name),"\\."),
          #      #                   function(x) paste(x[1:(length(x)-1)], collapse="."))
          #      #}
          #  }

            #if(substr(base,nchar(base),nchar(base))=="R"){
            if(single_rev){
                isolate({
                    files_info <<- paste0("fwd (F): ",fwd_file_name,"\nrev (R): ",rev_file_name," \n<em>aligned to: ",ref,"</em>",sep="")
                    #single_rev <- TRUE
                })
            }else if(is.null(ex)){
                isolate({
                    files_info <<- paste0("fwd (F): ",fwd_file_name,"\nrev (R): ",rev_file_name," \n<em>aligned to: ",ref,"</em>",sep="")
                })
            }
            if(load_id == 1){
                files_info  <<-  "demo file loaded\nframeshift deletion (c.277_278delCT) and heterozygous polymorphism (c.215C>G), detectable with these settings:\nmutation minimum peak % =~ 7; minimum quality =~ 16; 'use detected hetero indels' = checked"
                output$files <- renderPrint({cat("<pre>frameshift deletion (c.277_278delCT) and heterozygous polymorphism (c.215C>G), detectable with these settings:\nmutation minimum peak % =~ 7; minimum quality =~ 16; 'use detected hetero indels' = checked</pre>")})
                base = ""
                ex <- NULL
            }

            withProgress(message = paste('processing...',sep=" "), value = 1, {

                tryCatch(
                    g_abif <- sangerseqR::read.abif(fwd_file)@data,
                    error = function(e){output$files <- renderPrint(paste0("<pre>error while reading forward file, are you loading .abi ? ",e$message,"</pre>" ))})
                if(!is.null(rev_file)) {
                    tryCatch(
                        g_abif_rev <- sangerseqR::read.abif(rev_file)@data,
                        error = function(e){output$files <- renderPrint(paste0("<pre>error while reading reverse file, are you loading .abi ? ",e$message,"</pre>" ))})
                }
                else g_abif_rev <- NULL

                res <- NULL
                called <- NULL
                #res <- get_call_data(g_abif,g_abif_rev,input$rm7qual_thres,input$qual_thres,input$aln_min)
                tryCatch(

                    called <- suppressWarnings(get_call_data(g_abif,g_abif_rev,single_rev,g_glassed_ref)),
                    error = function(e){output$files <- renderPrint(paste0("<pre>error while loading calls from abi file : ",e$message,"</pre>" ))})

                if(!is.null(called)){
                    g_minor_het_insertions  <<- NULL
                    g_stored_het_indels     <<- list()
                    g_indels_present        <<- FALSE
                    g_qual_present          <<- called$qual_present
                    intensified  <-  get_intensities(g_abif,g_abif_rev,calls=called$calls,deletions=called$deletions,norm=FALSE,single_rev)
                    g_intens     <<- intensified$intens
                    g_intens_rev <<- intensified$intens_rev
                    calls        <-  annotate_calls(calls=intensified$calls,intens=intensified$intens,intens_rev=intensified$intens_rev,g_glassed_cod)
                    calls        <-  adjust_ref_mut(calls,g_intens_rev)
                    g_max_y      <<- max(c(max(g_intens[,list(A,C,G,T)]),if(is.null(g_intens_rev)) 0 else max(g_intens_rev[,list(A,C,G,T)])))
                    #intrex contains intesities coordinates of start and end of introns/exons with the sequence id (position in sequence coordinates)
                        intrexdat            <- list()
                        intrexdat$intrex     <- list()
                        intrexdat$intrex     <- setnames(calls[!is.na(exon_intron),list(max(id)-min(id)+1,min(trace_peak),max(trace_peak)),by = exon_intron],c("attr","length","trace_peak","end"))
                        intrexdat$intrex     <- setnames(merge(intrexdat$intrex,calls[,list(id,trace_peak)],by="trace_peak"),"trace_peak","start")
                        intrexdat$max_x      <- max(c(nrow(g_intens),nrow(g_intens_rev))) # these numbers should be the same
                        intrexdat$new_sample <- TRUE
                    g_intrexdat       <<- splice_variants(intrexdat)
                    calls             <-  data.table(calls,key="id")
                    g_noisy_neighbors <<- get_noisy_neighbors(calls)
                    if(!called$qual_present){
                        files_info <- paste0(files_info,HTML("\n<strong style=\"color: red;\">no Phred qualities!</strong>"))
                    }
                    files_info <- paste0("<pre>",files_info,"</pre>")
                    output$files      <-  renderPrint({cat(files_info)})
                    g_new_sample      <<- TRUE
                    if(g_files[loaded==TRUE,]$brush_fwd ==0){
                        g_brush_fwd        <<- calls[call!="-",][25]$trace_peak
                    }else{
                        g_brush_fwd <<- g_files[loaded==TRUE,]$brush_fwd
                    }
                    if(!is.null(g_abif_rev)){
                        if(g_files[loaded==TRUE,]$brush_rev ==0){
                            g_brush_rev        <<- calls[nrow(calls[call_rev!="-",])-25]$trace_peak
                        }else{
                            g_brush_rev <<- g_files[loaded==TRUE,]$brush_rev
                        }
                    }
                    if(g_single_rev){
                        g_brush_fwd <<- calls[nrow(calls[call!="-",])-25]$trace_peak

                    }
                    updateTabsetPanel(session,'tabs',selected = "main")

                    if(nrow(g_files[loaded==T,]) == 1){                  #g_calls saved from previous session we test if they are compatible to reload
                        if(g_files[loaded==T,]$calls != ""){
                            load(g_files[loaded==T]$calls)
                            if(nrow(g_calls)==nrow(calls))
                                calls <- g_calls
                        }
                    }
                } else return(structure("error_reading_Rbin",class = "my_UI_exception"))
            })
        }
        g_stored_het_indels <<- list()
        g_calls <<- NULL
        return(calls)
    })

    varcall <- reactive({
        if(class(loading_processed_files())[1] != "my_UI_exception") {
            update_chosen_variants()
            goReset_handler()
            g_reactval$updateVar
            #goBrush_fw()

            calls<-loading_processed_files()
            if(is.null(g_calls)){
                g_calls <<- calls
            }
            #remove added minor het ins
            if(exists("g_minor_het_insertions") && !is.null(g_minor_het_insertions$added)){
                for(i in nrow(g_minor_het_insertions))
                {
                    added <- strsplit(g_minor_het_insertions[i]$added[[1]], split = " ")
                    g_calls <<- g_calls[!(id %in% added[[1]])]
                    ret <- remove_intensities(added,g_calls,g_intens,g_intens_rev,g_intrexdat,g_minor_het_insertions)
                    g_calls      <<- ret$calls
                    g_intens     <<- ret$intens
                    g_intens_rev <<- ret$intens_rev
                    g_intrexdat  <<- ret$intrexdat
                }
                g_minor_het_insertions[,added:=NULL]
                g_minor_het_insertions[,ins_added := NULL]
            }

            g_calls <<- call_variants(g_calls,input$qual_thres_to_call,input$mut_min,input$s2n_min,g_stored_het_indels,g_brush_fwd,g_brush_rev,input$incorporate_checkbox,g_single_rev)
            setkey(g_calls,id)

            report                  <- report_hetero_indels(g_calls)
            g_indels_present       <<- report$indels_present
            g_minor_het_insertions <<- report$minor_het_insertions
            g_hetero_indel_aln     <<- report$hetero_indel_aln
            g_hetero_ins_tab       <<- report$hetero_ins_tab
            g_hetero_del_tab       <<- report$hetero_del_tab
            g_hetero_indel_pid     <<- report$hetero_indel_pid
            g_hetero_indel_report  <<- report$hetero_indel_report

            if(input$incorporate_checkbox & g_indels_present){
                g_calls <<- incorporate_hetero_indels_func(g_calls,g_hetero_del_tab,g_hetero_ins_tab,g_minor_het_insertions)
            }
            setkey(g_calls,id)

            if(exists("g_minor_het_insertions") && !is.null(g_minor_het_insertions$added)){
                for(i in 1:nrow(g_minor_het_insertions)){
                    ret <- add_intensities(strsplit(g_minor_het_insertions[i]$added[[1]],split= " ")[[1]],g_calls,g_intens,g_intens_rev,g_intrexdat)
                }
                g_calls                          <<- ret$calls
                g_minor_het_insertions$ins_added <<- ret$ins_added
                g_intens                         <<- ret$intens
                g_intens_rev                     <<- ret$intens_rev
                g_intrexdat                      <<- ret$intrexdat
            }

            g_expected_het_indel <<- get_expected_het_indels(g_calls)
            g_calls   <<- retranslate(g_calls)
            g_choices <<- get_choices(g_calls,g_glassed_ref)
            return(TRUE)

        } else return(FALSE)
    })

    #
    # Render functions reacting to varcall
    #
    output$plot <- renderChromatography({
        if(varcall()) {
            g_intrexdat$max_y <- (g_max_y*100)/input$max_y_p
            ret<-chromatography(g_intens,g_intens_rev,g_single_rev,g_intrexdat,g_calls,g_choices,g_new_sample,g_noisy_neighbors,input$show_calls_checkbox,g_qual_present,g_brush_fwd,g_brush_rev)
            g_new_sample <<- FALSE
            return(ret)
        }
    })

    output$hetero_indel_pid <- renderPrint({
        if(varcall() ) {
            if(!is.null(g_expected_het_indel) && g_expected_het_indel[[1]] * 100 >= 1) het_indel_info <- paste0("expected starting at ~",g_expected_het_indel[[1]] * 100,"% ->\n")
            else                                                                       het_indel_info <- paste0(" ... none expected\n")
            cat(het_indel_info)
            cat(g_hetero_indel_report)
        }
    })

    output$infobox <- renderPrint({
        input$qual_thres_to_call
        #if(loading_processed_files() != "not") {
            if(input$choose_call_pos != "") {
                tryCatch({
                    if(!is.null(g_intens_rev)) {
                        cat(g_calls[id == input$choose_call_pos,paste0("pos ",id,"    ref ",reference,"   user ",user_sample,"    max.peak% ",round(sample_peak_pct,1),"\n ",exon_intron,"  @genomic ",gen_coord,"  @coding ",coding_seq,"  @codon ",codon,"\n\nF  mut ",mut_peak_base_fwd,"  \tQ ",quality_fwd,"  \tpeak% ",round(mut_peak_pct_fwd,digits=1),"  \tS/N ",round(mut_s2n_abs_fwd,digits=1),"\nR  mut ",mut_peak_base_rev,"  \tQ ",quality_rev,"  \tpeak% ",round(mut_peak_pct_rev,digits=1),"  \tS/N ",round(mut_s2n_abs_rev,digits=1),sep="")])
                    } else {
                        cat(g_calls[id == input$choose_call_pos,paste0("pos ",id,"    ref ",reference,"   user ",user_sample,"    max.peak% ",round(sample_peak_pct,1),"\n ",exon_intron,"  @genomic ",gen_coord,"  @coding ",coding_seq,"  @codon ",codon,"\n\n   mut ",mut_peak_base_fwd,"  \tQ ",quality,    "  \tpeak% ",round(mut_peak_pct_fwd,digits=1),"  \tS/N ",round(mut_s2n_abs_fwd,digits=1),sep="")])
                    }
                }, error = function(er){
                    if(grepl("NAs introduced",er)) cat("type an integer number")
                })
                session$sendCustomMessage(type = 'input_change',message = paste0(g_calls[id==input$choose_call_pos]$trace_peak))
            } else cat("")
        #} else cat("load .abi/.ab1 file")  #!todo write this message somehwere else
    })

    #
    # Change single / Change button
    #
    update_chosen_variants <- reactive({
        input$change_btn
        isolate({
            if(!is.null(g_calls[["id"]])){
                if (input$change_user_sample != "") g_calls[id==as.numeric(input$choose_call_pos)]$user_sample <<- input$change_user_sample
                if (input$change_user_mut    != "") g_calls[id==as.numeric(input$choose_call_pos)]$user_mut    <<- input$change_user_mut
                # g_calls[id==as.numeric(input$choose_call_pos)]$user_variant <<- input$change_variant
                g_calls[id==as.numeric(input$choose_call_pos)]$set_by_user <<- TRUE
            }
        })
        return(T)
    })

    #
    # DataTable stuff
    #
#    add_checkboxes <- function(){
#        checkboxes <- paste0('<input type="checkbox" name="row', g_view$id, '" value="', g_view$id, '"',"")
#        for(i in 1:nrow(g_view)) {
#            if(g_view[i]$id %in% g_selected) checkboxes[i] <- paste0(checkboxes[i]," checked>","")
#            else checkboxes[i] <- paste0(checkboxes[i],">","")
#        }
#        return(checkboxes)
#    }

    output$chosen_variants_table <- DT::renderDataTable({
        if(varcall())
        if(!is.null(g_choices) && nrow(g_choices) > 0){
            input$change_btn
            #input$reset_btn
            #input$lo
            if(varcall() & !is.null(g_choices)) {
                g_view<<-get_view(g_calls,g_choices,g_glassed_snp)
                add_goto_buttons     <- shinyInput(actionButton, g_view$id, 'button_', label = "go",   onclick = 'Shiny.onInputChange(\"goGoto\",  this.id+ (Math.random()/10))' )
                add_reset_buttons    <- shinyInput(actionButton, g_view$id, 'button_', label = NULL, ico=rep("close",nrow(g_view)),onclick = 'Shiny.onInputChange(\"goReset\",  this.id)',class="btn dlt_btn" )
                add_lock_buttons     <- shinyInput(actionButton, g_view$id, 'button_', label = NULL,   onclick = 'console.log($("#DataTables_Table_1"));Shiny.onInputChange(\"goLock\",  this.id+ (Math.random()/10));if($(this).children(":first").attr("class")=="fa fa-unlock"){$(this).children().addClass(\'fa-lock\').removeClass(\'fa-unlock\');}else{$(this).children().addClass(\'fa-unlock\').removeClass(\'fa-lock\');}',ico = unlist(lapply(g_view$set_by_user, function(x){if(isTRUE(x)){"lock"}else{ "unlock"}})),class="btn btn-success" )
                out<-cbind(" "=add_goto_buttons, " "=add_reset_buttons, "<div title='Confirmed variants are kept for the session even if you change parameters,\nand appear in the samples panel from where they can be exported with the green export button.'>confirm [?]</div>"=add_lock_buttons, g_view[,list("call position (start)"=id,"genomic coordinate"=gen_coord,"coding variant"=coding,"protein variant"=protein,"pri peak %"=sample_peak_pct,"sec peak %"=mut_peak_pct)])
                if(!is.null(g_glassed_snp)){
                    out <- cbind(out,g_view[,list("dbSNP"=dbSNP)])
                }
                tableout<-DT::datatable(out
                                        , escape=FALSE
                                        #, class = "compact"
                                        , selection = "none"
                                        , style= 'bootstrap'
                                        , options=list("paging"=FALSE,"searching"=FALSE,"ordering"=FALSE,"autoWidth"=FALSE)
                                        )
            }
        }
    }
    #, options = list(dom = "t",orderClasses=c(-1,-2,-3,-4), paging=F, columnDefs=list(list(targets=c("_all"), searchable=F),list(targets=c(0,1,2,3), orderable=F, title="")))

    )

#    variant_select <- observe({
#        if(varcall()) {
#            g_selected <<- str_trim(input$rows)
#            g_selected_goto_index <<- 0
#        }
#    })

    #
    # data table buttons
    #
    goGoto_handler <- observe({
        if(is.null(input$goGoto)) return()
        goto_id <- floor(as.numeric(strsplit(input$goGoto, "_")[[1]][2]))/10
        session$sendCustomMessage(type = 'goto',message = paste0(g_calls[id==goto_id]$trace_peak))
        updateTextInput(session,"choose_call_pos",value=paste0(goto_id))
    })

    goReset_handler <- reactive({
        #if(varcall()) {
            if(!is.null(input$goReset)){
                isolate({
                    reset_id <- as.numeric(strsplit(input$goReset, "_")[[1]][2])
                    if(!is.na(g_view[id==reset_id]$ids)){
                        ids <- strsplit(g_view[id==reset_id]$ids," ")[[1]]
                        for( cid in ids){
                            updateTextInput(session,"choose_call_pos",value=paste0(cid))
                            g_calls[id==cid]$user_sample <<- g_calls[id==cid]$reference
                            g_calls[id==cid]$user_mut    <<- g_calls[id==cid]$reference
                            g_calls[id==cid]$set_by_user <<- TRUE
                        }
                    }else{
                        updateTextInput(session,"choose_call_pos",value=paste0(reset_id))
                        #reset button changed to remove variant
                        g_calls[id==reset_id]$user_sample <<- g_calls[id==reset_id]$reference
                        g_calls[id==reset_id]$user_mut    <<- g_calls[id==reset_id]$reference
                        g_calls[id==reset_id]$set_by_user <<- TRUE
                    }
                })
            }
            return(T)
        #}
    })

    goLock_handler <- observe({
        if(is.null(input$goLock)) return()
        isolate({
            lock_id <- strsplit(input$goLock, "_")[[1]][2]

            if(length(strsplit(lock_id,".",fixed=TRUE)[[1]]) >2){
                lock_id <- strsplit(lock_id,".",fixed=TRUE)[[1]]
                lock_id <- as.numeric(paste0(lock_id[1],".", lock_id[2]))
            }else{
                lock_id <- floor(as.numeric(lock_id))/10
            }


            g_view[id==lock_id]$set_by_user <<- !g_view[id==lock_id]$set_by_user
            coding  <- g_view[id==lock_id]$coding
            updateTextInput(session,"choose_call_pos",value=paste0(lock_id))
                    if(length(grep("ins|del|dup",coding)) > 0){ #locking indels
                        len <- length(unique(na.omit(as.numeric(unlist(strsplit(unlist(coding), "[^0-9]+"))))))
                        if(is.null(g_stored_het_indels[[coding]])){
                            g_stored_het_indels[[coding]] <<- lock_id
                            for(i in 0:(len-1)){
                                g_calls[id==(lock_id + i)]$set_by_user <<- !g_calls[id==(lock_id +i)]$set_by_user
                            }
                        }else{
                            g_stored_het_indels[[coding]] <<- NULL
                            for(i in 0:(len-1)){
                                g_calls[id==(lock_id+i)]$set_by_user <<- !g_calls[id==(lock_id +i)]$set_by_user
                            }
                        }

                    } else {     #locking SNPs
                g_calls[id==lock_id]$set_by_user <<- !g_calls[id==lock_id]$set_by_user
                abc<-123
            }

            if(nrow(g_view[set_by_user==TRUE])>0){
                big_space <- "&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp"
                prots <- paste(g_view[set_by_user == TRUE]$protein,collapse="")
                if(prots != ""){
                    prots <- paste0("<br>",big_space,"(",paste(g_view[set_by_user == TRUE]$protein,collapse=";"),")")
                }
                g_files<<-g_files[loaded==TRUE,status:=paste0("<b>confirmed</b>: ",paste(g_view[set_by_user == TRUE]$coding,collapse=";"),prots)]
                # shinyjs::show("export_btn")
            }else{
                g_files<<-g_files[loaded==TRUE,status:="viewed"]
                # shinyjs::show("export_btn")
            }
            #output$goLock <- renderUI({actionButton(input$goLock, icon = icon("lock"))})
        })
        return(T)
    })

    #
    # Input binding from JS
    #
    goClick_handler <- observe({
        if(is.null(input$pos_click)) return()
        isolate({
            updateTextInput(session,"choose_call_pos",value=paste0(input$pos_click$id))
        })
    })

    goBrush_fw <- observe({
        if(is.null(input$brush_fw)) return()
        last <- g_brush_fwd
        g_brush_fwd <<- input$brush_fw$coord
        #print(input$brush_fw$coord)
        #print(g_brush_fwd)
        if(last <= g_brush_fwd){
            if(nrow(g_choices[trace_peak > last][trace_peak < g_brush_fwd])==0)
                return()
        }
        g_reactval$updateVar <- runif(1,0,1)
    })
    goBrush_rv <- observe({
        if(is.null(input$brush_rv)) return()
        last <- g_brush_rev
        g_brush_rev <<- input$brush_rv$coord
        if(g_single_rev){     #we use brush_fwd (as we only have forward calls) but visualised and logic as brush rev
            g_brush_rev <- last
            g_brush_fwd <<- input$brush_rv$coord
        }else{
            if(last>=g_brush_rev){
                if(nrow(g_choices[trace_peak_rev < last][trace_peak_rev>g_brush_rev])==0)
                   return()
            }
        }
        g_reactval$updateVar <- runif(1,0,1)
    })


    #
    # Send message to JS
    #
    s2n_slider_handler <- observe({
            session$sendCustomMessage(type = 's2n_min',message = paste0(input$s2n_min))
    })


    join_traces <- observe({
        loading_processed_files()
        if(is.null(g_intens_rev))   session$sendCustomMessage(type = "join",message = "TRUE")
        else                        session$sendCustomMessage(type = "join",message = paste0(input$join_traces_checkbox))
    })

#     show_calls <- observe({
#         if(varcall()){
#             session$sendCustomMessage(type = "show",message = paste0(input$show_calls_checkbox))
#         }
#     })

    set_opacity <- observe({
        if(varcall()){
            opac_fwd <- 1 + (input$opacity/100)
            opac_rev <- 1 - (input$opacity/100)
            if(opac_fwd>1)opac_fwd <- 1
            if(opac_rev>1)opac_rev <- 1
            session$sendCustomMessage(type = "opac_f",message = paste0(opac_fwd))
            session$sendCustomMessage(type = "opac_r",message = paste0(opac_rev))
        }
    })


    #EXPORT

    output$export_btn <- downloadHandler(
        filename = function() {
            paste('data-', Sys.Date(), '.xlsx', sep='')
        },
        content = function(con) {
            out<-data.table("sample"=character(),"coding variant"=character(),"protein variant"=character())

            for(i in 1:nrow(g_files)){
                if(g_files[i]$status=="new"){
                    var = "NA"
                }else if(g_files[i]$status == "viewed"){
                    var = "wt"
                }
                names<-gsub("<b>confirmed</b>:","",g_files[i]$status)
                names<-gsub("&nbsp","",names)
                names<-strsplit(names,"<br>")[[1]]
                coding <- names[1]
                if(!is.na(names[2])) protein <- names[2]
                else protein<-""
                out <- rbind(out,list(paste0(g_files[i]$FWD_name,":",g_files[i]$REV_name),coding, gsub(")","",gsub('(',"",protein,fixed=TRUE),fixed=TRUE)))
            }
                #out<-rbind(out,list("1","2","2"))


            #g_selected <- g_view$id[as.numeric(input$chosen_variants_table_rows_selected)]

            #for(i in 1:nrow(g_view)) {
            #    if(g_view[i]$id %in% g_selected) out<-rbind(out,g_view[i,list("genomic coordinate"=gen_coord,"coding variant"=coding,"protein variant"=protein)])
            #}
            write.xlsx(out,con)
        }
    )

    #
    # Other tabs
    #

#    output$aln <- renderPrint({
#      if(varcall() ) {
#          cat("(P)rimary vs (S)econdary (or consensus of fwd+rev secondaries)\n\nidentified insertions:\n")
#          if(is.na(g_hetero_ins_tab[1])) cat("no insertions\n")
#          else print(g_hetero_ins_tab)
#          cat("\nidentified deletions:\n")
#          if(is.na(g_hetero_del_tab[1])) cat("no deletions\n")
#          else print(g_hetero_del_tab)
#          cat("\n")
#          writePairwiseAlignments(g_hetero_indel_aln, block.width = 150)
#      }
#    })
#
#    output$het_histogram <- renderPlot({
#        if(varcall() ) {
#            if(!is.null(g_expected_het_indel)){
#                plot(g_expected_het_indel$hist)
#                if(g_expected_het_indel$min != 0) abline(v = g_expected_het_indel$min,col = "red")
#                #if(g_expected_het_indel$max != 0) abline(v = g_expected_het_indel$max,col = "orange")
#            }
#        }
#    },height = 600)
#
#     output$call_table <- shiny::renderDataTable({
#         if(varcall() & !is.null(g_calls)) { g_calls }
#     }
#     ,options = list(paging=F
#                     ,columnDefs=list(list(searchable=F, orderable=F, title=""))
#                     ))
#
#     output$intens_table <- shiny::renderDataTable({
#         if(varcall() & !is.null(g_intens)) { g_intens }
#     }, options = list(paging=T, columnDefs=list(list(searchable=F, orderable=F, title=""))))

#     output$intens_table_rev <- shiny::renderDataTable({
#         if(varcall() & !is.null(g_intens_rev)) { g_intens_rev }
#     }, options = list(paging=T, columnDefs=list(list(searchable=F, orderable=F, title=""))))

    #used for conditional display
    output$reverse <- reactive({
        loading_processed_files()
        return(!is.null(g_intens_rev))
    })

    output$indels_present <- reactive({
        varcall()
        return(g_indels_present)
    })
    #output$show_sample_brows <- reactive({
    #    input$mng_samples_btn
    #    return(FALSE)
    #})
    #not shure why but I need this here to make it work (regirsters the output variable?)
    outputOptions(output, 'reverse', suspendWhenHidden=FALSE)
    outputOptions(output, 'indels_present', suspendWhenHidden=FALSE)
    #outputOptions(output, 'show_sample_brows',suspendWhenHidden=FALSE)
})
